programs/bpf_loader/src/lib.rs

```diff
@@ -61,70 +61,74 @@ pub enum BpfError {
-    e: EbpfError<BpfError>,
-) -> InstructionError {
+fn map_ebpf_error(invoke_context: &dyn InvokeContext, e: EbpfError<BpfError>) -> InstructionError {
     ic_msg!(invoke_context, "{}", e);
     InstructionError::InvalidAccountData
 }
 
-pub fn create_and_cache_executor(
-    key: &Pubkey,
-    data: &[u8],
+pub fn create_executor(
+    program_account_index: usize,
+    program_data_offset: usize,
     invoke_context: &mut dyn InvokeContext,
     use_jit: bool,
 ) -> Result<Arc<BpfExecutor>, InstructionError> {
+    let syscall_registry = syscalls::register_syscalls(invoke_context).map_err(|e| {
+        ic_msg!(invoke_context, "Failed to register syscalls: {}", e);
+        InstructionError::ProgramEnvironmentSetupFailure
+    })?;
     let bpf_compute_budget = invoke_context.get_bpf_compute_budget();
-    let mut program = Executable::<BpfError, ThisInstructionMeter>::from_elf(
-        data,
-        None,
-        Config {
-            max_call_depth: bpf_compute_budget.max_call_depth,
-            stack_frame_size: bpf_compute_budget.stack_frame_size,
-            enable_instruction_meter: true,
-            enable_instruction_tracing: log_enabled!(Trace),
-        },
-    )
+    let config = Config {
+        max_call_depth: bpf_compute_budget.max_call_depth,
+        stack_frame_size: bpf_compute_budget.stack_frame_size,
+        enable_instruction_meter: true,
+        enable_instruction_tracing: log_enabled!(Trace),
+    };
+    let mut executable = {
+        let keyed_accounts = invoke_context.get_keyed_accounts()?;
+        let program = keyed_account_at_index(keyed_accounts, program_account_index)?;
+        let account = program.try_account_ref()?;
+        let data = &account.data()[program_data_offset..];
+        Executable::<BpfError, ThisInstructionMeter>::from_elf(data, None, config)
+    }
     .map_err(|e| map_ebpf_error(invoke_context, e))?;
-    let (_, elf_bytes) = program
+    let (_, elf_bytes) = executable
         .get_text_bytes()
         .map_err(|e| map_ebpf_error(invoke_context, e))?;
     bpf_verifier::check(elf_bytes)
         .map_err(|e| map_ebpf_error(invoke_context, EbpfError::UserError(e)))?;
-    let syscall_registry = syscalls::register_syscalls(invoke_context).map_err(|e| {
-        ic_msg!(invoke_context, "Failed to register syscalls: {}", e);
-        InstructionError::ProgramEnvironmentSetupFailure
-    })?;
-    program.set_syscall_registry(syscall_registry);
+    executable.set_syscall_registry(syscall_registry);
     if use_jit {
-        if let Err(err) = program.jit_compile() {
+        if let Err(err) = executable.jit_compile() {
             ic_msg!(invoke_context, "Failed to compile program {:?}", err);
             return Err(InstructionError::ProgramFailedToCompile);
         }
     }
-    let executor = Arc::new(BpfExecutor { program });
-    invoke_context.add_executor(key, executor.clone());
-    Ok(executor)
+    Ok(Arc::new(BpfExecutor {
+        program: executable,
```

Comment:
nit, change the BpfExecutor member name to `executable` too?

Also, would be helpful to the reviewer to move the naming changes out and mainline them in a different pr